primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -0,0 +1,537 @@
+        // in that case compute the inverses individually
+        if accum_prod == P::Fr::zero() {
+            for i in 0..(vec_state.len() - 1) {
+                if vec_state[i][0] != P::Fr::zero() {
+                    vec_state[i][0] = vec_state[i][0].inverse().unwrap();
+                }
+            }
+        } else {
+
+            // Calculate the inversion of the products
+            // Use Montgomery simultaneous inversion
+            let mut w_bar = accum_prod.inverse().unwrap();
+
+            // Extract the individual inversions
+            let mut idx: i64 = w.len() as i64 - P::R as i64;
+            for i in (0..vec_state.len()).rev() {
+                let vec_1 = vec_state[i][0].clone();
+                vec_state[i][0] = w_bar * &w[idx as usize];
+                w_bar = w_bar * &vec_1;
+                idx -= 1;
+            }
+        }
+    }
+
+    pub fn poseidon_perm_gen(vec_state: &mut [Vec<P::Fr>]) {
+
+        // index that goes over the round constants
+        let mut round_cst_idx: usize = 0;
+
+        // Full rounds
+        for _i in 0..P::R_F {
+            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+
+        }
+
+        // Partial rounds
+        for _i in 0..P::R_P {
+            Self::poseidon_partial_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+        }
+
+        // Full rounds
+        // Last round does not contain the matrix mix
+        for _i in 0..P::R_F {
+            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
+
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+        }
+    }
+}
+
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr = F>> BatchFieldBasedHash for PoseidonBatchHash<F, P> {
+    type Data = F;
+    type BaseHash = PoseidonHash<F, P>;
+
+    fn batch_evaluate(input_array: &[F]) -> Result<Vec<F>, Error> {
+
+        // Input:
+        // This function calculates the hashes of inputs by groups of the rate P::R.
+        // The inputs are arranged in an array and arranged as consecutive chunks
+        // Example:
+        // For P::R = 2,
+        // (d_00, d_01, d_10, d_11, d_20, d_21, ...
+        // where d_00 and d_01 are the inputs to the first hash
+        // d_10 and d_11 are the inputs to the second hash ... etc..
+        // Output:
+        // The output is returned as an array of size input length / P::R
+
+        use rayon::prelude::*;
+
+        // Checks that size of input/output vector
+        assert_eq!(P::T - P::R, 1, "The assumption that the capacity is one field element is not satisfied.");
```

Comment:
Why is this strange assertion done? It depends on the targeted security level and the field size whether one field element is enough for the capacity of the sponge. For example, to yield 128 bit collision resistance at a field size of 256 bit, 1 field element capacity is enough. However if one targets 256 bit collision resistance, we would need 2 field elements capacity.

Reply:
I think it addresses the fact that this PoseidonHash implementation is for a 2:1 hash. If it's not related to this topic then we can remove it